README.md
isodumper
=========

A tool for writing ISO images on a USB stick. It's a fork of usb-imagewriter. 
We added also a formatting feature and a backup of USB device content.


Requirements
-------------

- coreutils
- pango
- polkit
- procps (or procps-ng)
- pygtk2.0-libglade
- python
- python-parted
- udisks
- xterm


License
--------

This software is distributed under the terms of the
[GNU General Public License version 2 (GPLv2+)](COPYING.GPL).

The [isodumper icon](isodumper.svg) is licensed under the terms of the
[GNU Lesser General Public License version 2.1 (LGPLv2.1+)](COPYING.LGPL).
It is part of the Crystal Clear icon set by Everaldo Coelho.

Authors
--------
- Papoteur, <papoteur@mageialinux-online.org>


Contributors
--------
- david.david, <david.david@mageialinux-online.org>


Translators
--------
- Papoteur, Guillaume Bernard, Eric Barbero, Rémi Verschelde - French
- Sergei Zivukov, AlexL, Анатолий Валерианович, Valentin XliN Saikov - Russian
- Yuri Chornoivan - Ukrainian
- marja - Dutch
- motitos, Miguel Ortega, Rodrigo Prieto, Egoitz Rodriguez Obieta - Spanish
- Marcello Anni - Italian
- Atilla ÖNTAŞ, tarakbumba, Fırat Kutlu - Turkish
- Filip Komar - Slovenian
- Kristoffer Grundström, Michael Eklund - Swedish
- You-Cheng Hsieh - Chinese Taiwan
- Marcio Andre Padula - Brazilian Portuguese
- Florin Cătălin RUSSEN - Romanian
- kiki.syahadat - Indonesian
- Daniel Napora, Tatsunoko - Polish
- Oliver Grawert, Rémi Verschelde, Marc Lattemann, user7, Christian Ruesch, psyca - German
- Marek Laane - Estonian
- dtsiamasiotis, Dimitrios Glentadakis - Greek
- fri, Jiří Vírava - Czech
- Rémi Verschelde - Danish
- richardshaylo, MrsB - English (United Kingdom)
- Manuela Silva - Portuguese
- Ḷḷumex03 - Asturian
- Egoitz Rodriguez Obieta - Basque
- Ardit Dani - Albanian


CHANGELOG
The initial state comes from https://launchpad.net/usb-imagewriter
The modifications are:

IsoDumper 0.32
------------
	  - fix wrong progress steps with small images.
	  - correction of the index not getting 100%.
	  - suppression of an inappriopriate emergency dialog after format.
	  - not closing the application after the success writing.
	  - added a dialog box to confirm overwriting backup file.
	  - adding the path in the destination name for backup file.
	  - change the name in .img for backup.
	  - default folder to /home/user for image filepicker.
	  - added a new translation on Albanian. (Thanks Ardit Dani)
	  - formatting and updating all .po files.
	  - recompiling all .mo files.

IsoDumper 0.31
------------
	  - this is a security update, about writing log file:
	  - see https://bugs.mageia.org/show_bug.cgi?id=14207.
	  - fix security issue when writing log file.

IsoDumper 0.30
------------
	  - added function to format the device in Fat32, ntfs or ext4.
	  - separated backup from writing.
	  - added a dialog box to warning against close when operationin action.
	  - added close the application when X of the window bar clicked.
	  - added file name and device name in the logview when selected.
	  - changed a window to a dialog form.
	  - set the user owner of log file and directory.
	  - suppress window redraw when closing the logview.
	  - added a new translation on Asturian. (Thanks Ḷḷumex03)
	  - added a new translation on Basque. (Thanks Egoitz Rodriguez Obieta)

IsoDumper 0.23
------------
	  - added close the application when X of the window bar clicked.
	  - added file name and device name in the logview when selected.
	  - set the user owner of log file and directory.
	  - suppress window redraw when closing the logview.
	  - fix some libglade-WARNING by deletion of swapped parameter.

IsoDumper 0.22
------------
	  - changed a window in a dialog form.
	  - added default values for the backup name and directory.
	  - improvement to the isodumper icon with a new SVG.
	  - improvement to the Makefile.

IsoDumper 0.21
------------
	  - added an "About isodumper" window "GtkAboutDialog".
	  - Integration of find_devices in the main file.
	  - fix of the behaviour of windows, not closing the application after escape from the confirmation dialog.
	  - fix transmission of the user name to write the log file at the good place.
	  - update all translations for new changes.

IsoDumper 0.20
------------
	  - added function to backup the content of the device before overwriting.
	  - added Release number in the window title (on isodumper.py).
	  - use mageia polkit policy instead of usermode. (Thanks to Colin Guthrie)
	  - replace all windows icons to icon isodumper.
	  - added and update Makefile for polkit file translation. (Thanks to Yuri Chornoivan)
	  - improvement po-update.sh script for translations (extract or merge polkit and desktop files translations).
	  - fix an error space on a translation string (#: lib/isodumper.py:270)
	  - added and update Estonian, Danish, Czech, Greek and English (United Kingdom) translations.
	  - updated  de,fr,id,ro,sl,sv,zh_TW,tr,uk,pt_BR,ru,ro,es,it and pl translations.

IsoDumper 0.13
------------
	  - added and update Polish and German translations.
	  - updated Dutch and Italian translations.
	  - added a Makefile file for build package.
	  - added a pixmaps icons.
	  - usermode-consoleonly is not a BuildRequires, but it is a Requires package.
	  - improvement po-update.sh script for translations.
	  - added Release number in the window title.

IsoDumper 0.12
------------
	  - Correction of bugs in the device name and in the dialog box loops
	  - Added a catch of writing error

IsoDumper 0.11
------------
	  - added check for the size of the destination device.
	  - added a button to retry in the no_device dialog box.
	  - added a dialog box for devices with a size of 32 GB and more.
	  - added and update Swedish, Chinese, Brazilian Portuguese, Indonesian and Romanian translations.
	  - updated Ukrainian, Russian, Slovenian, Spanish, Turkish Translations.

IsoDumper 0.10
------------
	  - added and update Russian, Ukrainian, Spanish, Dutch, Italian, Turkish and Slovenian translations
	  - update French translation
	  - added ru, uk, es, nl, it, tr and sl comment in .desktop file
	  - added po-update.sh and po-compile.sh files for translations (see i18n.md for more details)


IsoDumper 0.1
------------
	  - changing the name into isodumper
	  - replaced the dd command with direct writing command python.
	  - progressbar is functionnal.
	  - replaced the banner (header.png) with an Mageia's header.
	  - added .iso filter.
	  - added writing ~/.isodumper/isodumper.log at the end.
	  - writing progression in details only each 1%.
	  - point at the end of the logview window.
	  - indicating the name of the log file.
	  - new feature : the application doesn't close when a stick is inserted after the start and a dialog box which is displayed to say it.


export.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun 25 19:00:13 2013

@author: yves
"""
import os
# Block size
    def raw_write(self, source, target):
        bs=4096
        b = os.path.getsize("/documents/boot-nonfree.iso")
        ifc=open(source, "rb",1)
        ofc= open("/documents/test.iso", 'wb',0)
        progress = self.wTree.get_widget("progressbar")
        progress.set_sensitive(True)
        progress.set_text(_('Writing ')+source.split('/')[-1]+_(' to ')+self.dev)
        self.logger(_('Executing copy from ')+source+' to '+target)
        while gtk.events_pending():
           gtk.main_iteration(True)
        #output = Popen(['dd if='+source+' of='+target+' bs=1024'], stdout=PIPE, stderr=STDOUT, shell=True)
        #self.ddpid = output.pid
        #while output.stdout.readline():
        #    line = output.stdout.readline().strip()
        steps=range(0, b, b/100)
        indice=1
        written=0
        for i in xrange(0,b,bs):
            try:
                buf=ifc.read(bs)
            except:
                self.logger(_("Reading error."))
                return False
            try:
                ofc.write(buf)
            except:
                self.logger(_("Writing error."))
                return False
            written= written+bs
            if written > steps[indice]:
                indice +=1
                while gtk.events_pending():
                    gtk.main_iteration(True)
            #if line.endswith('MB/s'):
            #    target_size = line.split()[0]
                self.logger(_('Wrote: ')+written+' bytes')
                #size = float(target_size)*100/float(src_size)
                while gtk.events_pending():
                    gtk.main_iteration(True)
                progress.set_fraction(float(indice/100))
        #pid, sts = os.waitpid(output.pid, 0)
        #if sts != 0:
        #    self.logger(_('The dd process ended with an error !'))
        #    self.emergency()
        #    return False
        progress.set_fraction(1.0)
        self.logger(_('Image ')+source.split('/')[-1]+_(' successfully written to')+target)
        self.logger(_("%s octets écrits")%written)
        self.success()

        # tous les 5%
            
        ifc.close()
        ofc.close()
        print 


install.sh
#!/bin/sh
#
#    Copyright 2007-2009 Canonical Ltd.
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License as
#  published by the Free Software Foundation; either version 2 of the
#  License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software Foundation,
#  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA

LIBFILES="isodumper.py find_devices"
DATAFILES="isodumper.glade header.png"


if [ "$1" = "uninstall" ]; then
    rm -rf /usr/lib/isodumper
    rm -rf /usr/share/isodumper
    rm -r /usr/share/applications/isodumper.desktop
    rm -f /usr/bin/isodumper
else
    cp share/applications/isodumper.desktop /usr/share/applications/
    cp isodumper /usr/bin/
    mkdir -p /usr/lib/isodumper
    mkdir -p /usr/share/isodumper

    for item in $LIBFILES; do
        cp lib/$item /usr/lib/isodumper/
    done

    for item in $DATAFILES; do
        cp share/usb-isodumper/$item /usr/share/isodumper/
    done
fi


tools/po-update.sh
#!/bin/sh

set -e

if test "$1" = "-h" -o "$1" = "--help"; then
	echo "Use: $0 [<language>]"
	echo "Run without arguments to update all translation files."
	exit 0
fi

cd "$(readlink -f "$(dirname "$0")/..")"

VERSION=(0.40)
DOMAIN=(isodumper)

intltool-extract --type=gettext/glade share/isodumper/isodumper.glade
intltool-extract --type=gettext/xml polkit/org.mageia.isodumper.policy.in
intltool-extract --type=gettext/ini share/applications/isodumper.desktop.in

POT_DIR="$PWD/po"
test -d "$POT_DIR"

POT_FILE="$POT_DIR/$DOMAIN.pot"

/usr/bin/xgettext \
	--package-name "$DOMAIN" \
	--package-version "$VERSION" \
	--language=Python --from-code=UTF-8 --keyword=_ --keyword=N_ \
	--no-escape --add-location --sort-by-file \
	--add-comments=I18N \
	--output="$POT_FILE" \
	lib/isodumper.py \
	share/isodumper/isodumper.glade.h \
	polkit/org.mageia.isodumper.policy.in.h \
	share/applications/isodumper.desktop.in.h

/bin/sed --in-place --expression="s/charset=CHARSET/charset=UTF-8/" "$POT_FILE"

intltool-merge --desktop-style po share/applications/isodumper.desktop.in share/applications/isodumper.desktop

rm -f share/isodumper/isodumper.glade.h
rm -f polkit/org.mageia.isodumper.policy.in.h
rm -f share/applications/isodumper.desktop.in.h

update_po() {
	local LL_CC="$1"
	local PO_FILE="$POT_DIR/$LL_CC.po"

        echo "Update $(basename "$PO_FILE"):"
	/usr/bin/msgmerge \
		--update --no-fuzzy-matching \
		--no-escape --add-location --sort-by-file \
		--lang="$LL_CC" \
		"$PO_FILE" "$POT_FILE"

	# /bin/sed --in-place --expression="s/Language: \\\\n/Language: $L_NAME\\\\n/" "$PO_FILE"
}

if test "$1"; then
	update_po "$1"
else
	for l in $(ls -1 "$POT_DIR"/*.po); do
		l="$(basename "$l")"
		update_po "${l%.po}"
	done
fi


lib/raw_format.py
#!/usr/bin/python
#
#  imported from project Mintstick, by Clement Lefebvre
#
#  Copyright (C) 2013 THE isodumper'S COPYRIGHT HOLDER
#  This file is distributed under the same license as the isodumper package.
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License as
#  published by the Free Software Foundation; either version 2 of the
#  License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software Foundation,
#  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
#
#  Author isodumper software= papoteur <papoteur@mageialinux-online.org>

import os, sys
import getopt
import parted
from subprocess import call

sys.path.append('/usr/lib/isodumper')

def do_umount(target):
        mounts = get_mounted(target)
        if mounts:
            print 'Unmounting all partitions of '+target+':'
        for mount in mounts:
            print 'Trying to unmount '+mount[0]+'...'
            try:
                retcode = call('umount '+mount[0], shell=True)
                if retcode < 0:
                    print 'Error, umount '+mount[0]+' was terminated by signal '+str(retcode)
                    sys.exit(6)
                else:
                    if retcode == 0:
                        print mount[0]+' successfully unmounted'
                    else:
                        print 'Error, umount '+mount[0]+' returned '+str(retcode)
                        sys.exit(6)
            except OSError, e:
                print 'Execution failed: '+str(e)
                sys.exit(6)


def get_mounted(target):
        try:
            lines = [line.strip("\n").split(" ") for line in open ("/etc/mtab", "r").readlines()]
            return [mount for mount in lines if mount[0].startswith(target)]
        except:
             print 'Could not read mtab !'
             sys.exit(6)

def raw_format(device_path, fstype, volume_label, uid, gid):

    do_umount(device_path)

    # First erase MBR and partition table , if any
    os.system ("dd if=/dev/zero of=%s bs=512 count=1 >/dev/null 2>&1" % device_path)

    device = parted.getDevice(device_path)

    # Create a default partition set up
    disk = parted.freshDisk(device, 'msdos')
    disk.commit()
    regions = disk.getFreeSpaceRegions()

    if len(regions) > 0:
        #print "Build partition"
        # Define size
        start = parted.sizeToSectors(1, "MiB", device.sectorSize)
        #print "start %s" % start
        end = device.getLength() - start - 1024
        #print end

        # Alignment
        #cylinder = device.endSectorToCylinder(end)
        #end = device.endCylinderToSector(cylinder)
        #print end
        try:
            geometry = parted.Geometry(device=device, start=start, end=end)
        except:
            print "Geometry error - Can't create partition"
            sys.exit(5)

        # fstype
        fs = parted.FileSystem(type=fstype, geometry=geometry)

        # Create partition
        partition = parted.Partition(disk=disk, type=parted.PARTITION_NORMAL, geometry=geometry, fs=fs)
        constraint = parted.Constraint(exactGeom=geometry)
        disk.addPartition(partition=partition, constraint=constraint)
        disk.commit()

        # Format partition according to the fstype specified
        if fstype == "fat32":
            os.system("mkdosfs -F 32 -n \"%s\" %s >/dev/null 2>&1" % (volume_label, partition.path))
        if fstype == "ntfs":
            os.system("mkntfs -f -L \"%s\" %s >/dev/null 2>&1" % (volume_label, partition.path))
        elif fstype == "ext4":
            os.system("mkfs.ext4 -E root_owner=%s:%s -L \"%s\" %s >/dev/null 2>&1" % (uid, gid, volume_label, partition.path))
    sys.exit(0)





def main():
    # parse command line options
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hd:f:l:u:g:", ["help", "device=","filesystem=","label=","uid=","gid="])
    except getopt.error, msg:
        print msg
        print "for help use --help"
        sys.exit(2)

    for o, a in opts:
        if o in ("-h", "--help"):
            print "Usage: %s -d device -f filesystem -l volume_label\n"  % sys.argv[0]
            print "-d|--device          : device path"
            print "-f|--filesystem      : filesystem\n"
            print "-l|--label           : volume label\n"
            print "-u|--uid             : uid of user\n"
            print "-g|--gid             : gid of user\n"
            print "Example : %s -d /dev/sdj -f fat32 -l \"USB Stick\" -u 1000 -g 1000" % sys.argv[0]
            sys.exit(0)
        elif o in ("-d"):
            device = a
        elif o in ("-f"):
            if a not in [ "fat32", "ntfs", "ext4" ]:
                print "Specify fat32, ntfs or ext4"
                sys.exit(3)
            fstype = a
        elif o in ("-l"):
            label = a
        elif o in ("-u"):
            uid = a
        elif o in ("-g"):
            gid = a

    argc = len(sys.argv)
    if argc < 11:
      print "Too few arguments"
      print "for help use --help"
      exit(2)

    raw_format(device, fstype, label, uid, gid)

if __name__ == "__main__":
    main()
